src/shogun/mathematics/linalg/LinalgBackendViennacl.h

```diff
@@ -262,6 +272,46 @@ class LinalgBackendViennaCL : public LinalgBackendGPUBase
 			+ beta * b_gpu->data_matrix(b.num_rows, b.num_cols);
 	}
 
+	/** ViennaCL convolve/cross-correlation method. */
+	template <typename T>
+	void convolve_impl(SGMatrix<T>& X, SGMatrix<T>& W, SGMatrix<T>& Y,
+		bool flip=false, index_t stride_x=1, index_t stride_y=1)
+	{
+		GPUMemoryViennaCL<T>* X_gpu = cast_to_viennacl(X);
+		GPUMemoryViennaCL<T>* W_gpu = cast_to_viennacl(W);
+		GPUMemoryViennaCL<T>* Y_gpu = cast_to_viennacl(Y);
+
+		if (stride_x==1 && stride_y==1)
+		{
+			viennacl::ocl::kernel& kernel = generate_kernel_unity_stride<T>(
+				(W.num_cols-1)/2, (W.num_rows-1)/2, flip);
+
+			kernel.global_work_size(0, linalg::implementation::ocl::align_to_multiple_2d(Y.num_cols));
+			kernel.global_work_size(1, linalg::implementation::ocl::align_to_multiple_2d(Y.num_rows));
+
+			viennacl::ocl::enqueue(kernel(
+				X_gpu->data_matrix(X.num_rows, X.num_cols), cl_int(X.num_cols),
+				cl_int(X.num_rows), cl_int(X_gpu->m_offset),
+				W_gpu->data_matrix(W.num_rows, W.num_cols), cl_int(W_gpu->m_offset),
+				Y_gpu->data_matrix(Y.num_rows, Y.num_cols), cl_int(Y_gpu->m_offset)));
+		}
+		else
```

Comment:
Yes they are tested. I think the problem is the code `X.vcl_matrix()` is now like `X_gpu->data_matrix(X.num_rows, X.num_cols)`